### PR TITLE
Using storage aliases

### PR DESCRIPTION
**Pull request by g4s8**
#78 - using storage aliases for repo configuration. Refactored `RepoConfig` methods to be synchronous to avoid one more (5-th) callback parameter for constructing configuration. This feature helps to hide actual credentials from config and restrict user permissions to allow modify repository settings only without changing storage configuration (in case of Central).

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#194](https://github.com/artipie/artipie/issues/194) is assigned to @brastak[/z](https://www.0crat.com/u/brastak), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/41d57d3b-1ff0-465b-ab1f-2213e718510f); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; @brastak[/z](https://www.0crat.com/u/brastak) is not a member of this project yet, but they can request to join, as [§1](http://www.zerocracy.com/policy.html#1) explains; there will be no monetary reward for this job

File: src/main/java/com/artipie/RepoConfig.java [hunk 1]
```diff
@@ -90,58 +100,82 @@ public RepoConfig(final Key prefix, final Publisher<ByteBuffer> content) {
...
+     * @return Completion stage of yaml
+     */
+    @SuppressWarnings("PMD.ProhibitPublicStaticMethods")
+    public static CompletionStage<RepoConfig> fromPublisher(final StorageAliases storages,
+        final Key prefix, final Publisher<ByteBuffer> pub) {
+        return Flowable.fromPublisher(pub)
+            .reduce(
+                new StringBuilder(),
+                (acc, buf) -> acc.append(
+                    new String(new Remaining(buf).bytes(), StandardCharsets.UTF_8)
```
Comment (brastak) [inline src/main/java/com/artipie/RepoConfig.java:172]: @g4s8 It's better to use Concatenation here and create the String only when all bytes are read. ByteBuffer can ends at the middle of UTF-8 char (let's assume that buffer size = 4. When you try to read `1фыв2`, byte buffer will read `1ф` and the first byte of `ы`. It causes the resulted string will be like `1ф??в2`)

File: src/main/java/com/artipie/StorageAliases.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        return storage.exists(key).thenCompose(
+            found -> {
+                final CompletableFuture<StorageAliases> res;
+                if (found) {
+                    res = storage.value(key).thenCompose(
+                        pub -> Flowable.fromPublisher(pub)
+                            .reduce(
+                                new StringBuilder(),
+                                (acc, buf) -> acc.append(
+                                    new String(new Remaining(buf).bytes(), StandardCharsets.UTF_8)
```
Comment (brastak) [inline src/main/java/com/artipie/StorageAliases.java:74]: @g4s8 The same issue - create the string after reading all bytes

File: src/main/java/com/artipie/repo/FlatLayout.java [hunk 3]
```diff
@@ -103,13 +104,16 @@ public Slice resolve(final String path) {
                         final Slice slice;
                         if (exist) {
                             slice = new AsyncSlice(
-                                storage.value(key).thenCombine(
+                                StorageAliases.find(storage, key.parent().get()).thenCompose(
```
Comment (brastak) [inline src/main/java/com/artipie/repo/FlatLayout.java:107]: @g4s8 It seems that we should to think about caching. We do a lot of config readings on each request. Maybe create the issue?

**Comment by g4s8** [comment]
@brastak thanks fixed your comments and created new ticket #195

**Comment by brastak** [comment]
@g4s8 Thanks, approved

**Comment by 0crat** [comment]
Job was finished in 22 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @brastak[/z](https://www.0crat.com/u/brastak), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Pull request merged**